docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+* It operates on Object Storage API instead of local filesystem.
+* Operations are done from multiple processes that lack coordination.
+
+Moving from lock-based logic to coordination free and from strongly consistent local filesystem to potentially eventually 
+consistent remote simplified "filesystem" in form of Object Storage API, causes additional cases that we need to consider 
+in Thanos system, like:
+* Thanos sidecar or compactor crashes during the process of uploading the block. It uploaded index, 2 chunk files and crashed. How to ensure readers (e.g compactor, store gateway) will handle this gracefully?
+* Thanos compactor uploads compacted block and deletes source blocks. After next sync iteration it does not see a new block (read after write eventual consistency). It sees gap, wrongly plans next compaction and causes non-resolvable overlap.
+* Thanos compactor uploads compacted block and deletes source blocks. Thanos Store Gateway syncs every 3m so it missed that fact. Next query that hits store gateway tries to fetch deleted source blocks and fails.
+
+Currently, we have only basic safeguards against those in form of [`syncDelay`](https://github.com/improbable-eng/thanos/blob/bc088285a1b4bf464fdf2539e4b365b805874eed/pkg/compact/compact.go#L187). 
+That helps with operations between sidecar and compactor but does not fix inter-compactor logic in case of eventual consistent storage or partial upload failures.
+
+We also partially assumed strong consistency for object storage, because [GCS has strong consistency](https://cloud.google.com/storage/docs/consistency) and S3 has as it as well, 
+but with [some caveats](https://codeburst.io/quick-explanation-of-the-s3-consistency-model-6c9f325e3f82) that we met.
+
+However with increasing number of other providers (ceph, hdfs, COS, Azure, Swift) with no (or unknown) strong consistency 
+guarantee this assumption needs to go away.
+
+We have an explicit API for Series fetching in form of gRPC StoreAPI, but we lack the verbose API for other side: the object storage format with assumptions on behavior that will
+cover partial blocks, eventual consistency and syncing requirements, which is a goal for this proposal.
+
+## Proposal
+
+We propose defining those set of rules to solve A, B and C goals that will define the object storage format API (and how to operate with those objects):
+
+1. Block once uploaded is never edited (no file within ULID directory is removed, edited or added).
+1. Block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+   to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+   still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+1. TSDB `meta.json` file within block *have to be* uploaded as first one.
```

Comment:
Since we don't have strongly consistent, why  does it matter that it is the first to be uploaded?

Reply:
Very good point, I will explain it further with:

>  It matters even when we don't have strong consistency, because 
it helps us detect the case of `upload` crashed in between of upload. We only upload meta.json after successful uploads of rest of the files for the block.